src/geometry/euclidean.lean

```diff
@@ -683,4 +683,230 @@ begin
+through the given point is the `orthogonal_projection` of that point
+onto the subspace. -/
+lemma inter_eq_singleton_orthogonal_projection {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) (p : P) :
+  (s : set P) ∩ (mk' p s.direction.orthogonal) = {orthogonal_projection hn hc p} :=
+inter_eq_singleton_orthogonal_projection_fn hn hc p
+
+/-- The `orthogonal_projection` lies in the given subspace. -/
+lemma orthogonal_projection_mem {s : affine_subspace ℝ V P} (hn : (s : set P).nonempty)
+    (hc : is_complete (s.direction : set V)) (p : P) : orthogonal_projection hn hc p ∈ s :=
+orthogonal_projection_fn_mem hn hc p
+
+/-- The `orthogonal_projection` lies in the orthogonal subspace. -/
+lemma orthogonal_projection_mem_orthogonal {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) (p : P) :
+  orthogonal_projection hn hc p ∈ mk' p s.direction.orthogonal :=
+orthogonal_projection_fn_mem_orthogonal hn hc p
+
+/-- Subtracting a point in the given subspace from the
+`orthogonal_projection` produces a result in the direction of the
+given subspace. -/
+lemma orthogonal_projection_vsub_mem_direction {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {p1 : P} (p2 : P)
+    (hp1 : p1 ∈ s) :
+  orthogonal_projection hn hc p2 -ᵥ p1 ∈ s.direction :=
+vsub_mem_direction (orthogonal_projection_mem hn hc p2) hp1
+
+/-- Subtracting the `orthogonal_projection` from a point in the given
+subspace produces a result in the direction of the given subspace. -/
+lemma vsub_orthogonal_projection_mem_direction {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {p1 : P} (p2 : P)
+    (hp1 : p1 ∈ s) :
+  p1 -ᵥ orthogonal_projection hn hc p2 ∈ s.direction :=
+vsub_mem_direction hp1 (orthogonal_projection_mem hn hc p2)
+
+/-- A point equals its orthogonal projection if and only if it lies in
+the subspace. -/
+lemma orthogonal_projection_eq_self_iff {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) (p : P) :
+  orthogonal_projection hn hc p = p ↔ p ∈ s :=
+begin
+  split,
+  { exact λ h, h ▸ orthogonal_projection_mem hn hc p },
+  { intro h,
+    have hp : p ∈ ((s : set P) ∩ mk' p s.direction.orthogonal) := ⟨h, self_mem_mk' p _⟩,
+    rw [inter_eq_singleton_orthogonal_projection hn hc p, set.mem_singleton_iff] at hp,
+    exact hp.symm }
+end
+
+/-- The distance between a point and its orthogonal projection is
+nonzero if it does not lie in the subspace. -/
+lemma dist_orthogonal_projection_ne_zero_of_not_mem {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {p : P} (hp : p ∉ s) :
+  dist p (orthogonal_projection hn hc p) ≠ 0 :=
+begin
+  intro hz,
+  rw dist_eq_zero at hz,
+  symmetry' at hz,
+  rw orthogonal_projection_eq_self_iff hn hc _ at hz,
+  exact hp hz
+end
```

Comment:
I'd use a more general route:

```suggestion
/-- The distance to a point's orthogonal projection is 0 iff it lies in the subspace. -/
lemma dist_orthogonal_projection_eq_zero_iff {s : affine_subspace ℝ V P}
  (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {p : P} :
  dist p (orthogonal_projection hn hc p) = 0 ↔ p ∈ s :=
by rw [dist_comm, dist_eq_zero, orthogonal_projection_eq_self_iff]

/-- The distance between a point and its orthogonal projection is
nonzero if it does not lie in the subspace. -/
lemma dist_orthogonal_projection_ne_zero_of_not_mem {s : affine_subspace ℝ V P}
    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {p : P} (hp : p ∉ s) :
  dist p (orthogonal_projection hn hc p) ≠ 0 :=
mt (dist_orthogonal_projection_eq_zero_iff hn hc).mp hp
```

Reply:
Oof, it looks like GitHub did not understand the diff. I meant to insert the following lemma above `dist_orthogonal_projection_ne_zero_of_not_mem`:
```lean
/-- A point lies in a subspace if the distance to its orthogonal projection is 0. -/
lemma dist_orthogonal_projection_eq_zero_iff {s : affine_subspace ℝ V P}
  (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {p : P} :
  dist p (orthogonal_projection hn hc p) = 0 ↔ p ∈ s :=
by rw [dist_comm, dist_eq_zero, orthogonal_projection_eq_self_iff]
```

so that the proof of `dist_orthogonal_projection_ne_zero_of_not_mem` simplifies to:
```lean
mt (dist_orthogonal_projection_eq_zero_iff hn hc).mp hp
```